akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala

```diff
@@ -422,6 +425,10 @@ class ClusterSingletonManager(
     removed = removed filter { case (_, deadline) ⇒ deadline.hasTimeLeft }
   }
 
+  // for CoordinatedShutdown
+  val memberExitingProgress = Promise[Done]()
+  CoordinatedShutdown(context.system).addTask("cluster-pre-exiting")(() ⇒ memberExitingProgress.future)
```

Comment:
I was thinking that each task must be well behaved, i.e. it should itself add timeout behavior. Also recover to not fail the future if not a fatal problem.